### DISCUSSION TITLE
nrf24l01 pico-to-pico officaltest gets me succes=0 failures=16

### DISCUSSION THREAD

**Discussion by UmutDemirel06** [micropython/micropython#12383] [nRF] [answered]
as ı said in the title ı am trying to use two nrf24l01 to communicate pico-to-pico with the official test but it gives me succes=0 failures=16 as seen in the picture ı use two computers each connected to one of the Picos ı will put a picture of the wiring and the screenshots of the thonny interface
![image](https://github.com/micropython/micropython/assets/59196975/845a420f-db39-42f6-8969-21d2bc6168a6)
![20230906_153154](https://github.com/micropython/micropython/assets/59196975/1717c302-608d-42a5-af83-916920bb4715)
![20230906_153144](https://github.com/micropython/micropython/assets/59196975/b2c9233d-91c8-47f5-adbf-b08d3a2eefe4)

**Comment by shariltumin** [comment]
The SPI pins for rp2 in "nrf24l01test.py" are defined as follows:

```
spi = SPI(0, sck=Pin(2), mosi=Pin(3), miso=Pin(4))
```

However, machine.SPI(0) is mapped to different pins in the firmware.

```
MicroPython v1.20.0-242-g6a9db521e on 2023-06-30; Raspberry Pi Pico with RP2040
Type "help()" for more information.
>>> import sys
>>> import machine
>>> sys.platform
'rp2'
>>> spi = machine.SPI(0)
>>> spi
SPI(0, baudrate=992063, polarity=0, phase=0, bits=8, sck=18, mosi=19, miso=16)
```

You can try changing the SPI pins in "nrf24l01test.py" to
```
elif usys.platform == "rp2":  # Hardware SPI with explicit pin definitions
    spi = SPI(0, sck=Pin(18), mosi=Pin(19), miso=Pin(16))
    cfg = {"spi": spi, "csn": 5, "ce": 6}
```

and rewire your setup accordingly. I think this will give you the result you are looking for.

**Comment by peterhinch** [comment]
@shariltumin Three observations.
 1. The pins in the code are correct according to [the official pinout](https://datasheets.raspberrypi.com/pico/Pico-R3-A4-Pinout.pdf).
 2. I tested the script successfully.
 3. Your alternative pins are also correct according to the pinout. Both are valid.

@Hopesforgypsy 
There is no need to change wiring or code.

**Reply by UmutDemirel06** [reply]
could it be maybe because of faulty pico or nrf24l01?

**Reply by shariltumin** [reply]
Instead of rp2-picos, you can try two esp32 boards. I have tried the test on esp32-s2 boards with success.

**Reply by UmutDemirel06** [reply]
ı will try but it will take some time

**Reply by steveh127** [reply, marked as answer]
The problem isn't with the SPI connections it's with the power supply. Since your original query I've been experimenting with these modules and, like you was get very poor communications between the boards. Reading the NRF24L01 datasheet there is a section on the importance of a clean power supply. This includes making sure the power lines aren't interfered with by digital line such as the SPI wires. Adding a capacitor between ground and 3.3v as close to the chip as possible was also suggested. 

If you have the soldering skills soldering a 10uf capacitor across the ground and 3.3v pins of the module works wonders. I did find making sure that power and SPI connections were kept apart helped too. If you happen to have any 0805 SMD 10uf capacitors they fit very nicely between the module pins on top of the board.

An alternative that should work is to power the modules from a separate, clean, power supply.  Batteries should work though I haven't tried this.

Something that also works is to connect the modules to the VSYS output of the pico, this supplies 5v so is out of spec for the module and could well damage them, try at your own risk. I've tried it in a quick test and performance was good and the modules still work. If I was using these modules seriously I'd put a 3.3v regulator between VSYS and the module which would guarantee a clean power supply.

Nothing much to do with micropython, but might get you a working setup so you can start doing more than testing.

**Reply by peterhinch** [reply]
@steveh127 That is useful information.

I have found that NRF24L01+ modules vary greatly in quality. Sparkfun units work well. Some Chinese ones struggled to communicate within the same room. It's possible that these differences were down to onboard decoupling. I use Sparkfun exclusively.

**Reply by UmutDemirel06** [reply]
thank you all for helping me get this at least working so that ı can move on to more involved projects

**Reply by steveh127** [reply]
The modules I've been testing recently where some I had lying around. Certainly not Sparkfun and probably cheap Chinese ones. I didn't test the range, they eventually worked OK but where inches apart. They were certainly the same size / pinout as Hopesforgegypsy's boards. I guess you get what you pay for. The Sparkfun boards are listed on Pi Hat at £22.30, on Amazon you can get 10 of the cheaper ones for half that.

**Reply by peterhinch** [reply]
I've just checked and the Sparkfun board is [retired](https://www.sparkfun.com/products/retired/705). So I guess we have no choice than to buy Chinese ones and get soldering. Unless there is another properly designed board?

As a very general point ESPNow on ESP32-S3 offers similar high-integrity communication with excellent micro-power sleep modes. I'm doubtful whether I would use NRF24L01 on new projects.

**Reply by steveh127** [reply]
pi Hut UK have a few Sparkfun boards in stock still. And, though retired, Sparkfun provide a schematic for their boards and the only real difference is the addition of a 3.3v regulator to ensure a clean power supply. However, like Peter I can't see a practical use for these given the alternatives.

**Reply by peterhinch** [reply]
There may be differences on the RF side. There were two Sparkfun boards, one with a reverse SMA connector and one with a chip antenna. The Chinese ones have a PCB trace antenna which may or may not be well designed. Then there is the quality and location of decoupling caps.

**Comment by shariltumin** [comment]
For those who may read this thread later. 

I have a device set up as a mobile transmitter, powered by a USB power bank. I have this in the 'main.py' file.

```
import mini_test as test
test.sender()
```

The other device is a listener and is connected to REPL so I can see what is coming.

The 'mini_test.py' test script:

```
# usage:-
# import mini_test as test
# test.sender() --or-- test.listener()

import struct, time, sys
from machine import Pin, SPI, SoftSPI
from nrf24l01 import NRF24L01

# depend on your board!
csn = Pin(5, mode=Pin.OUT, value=1)
ce = Pin(3, mode=Pin.OUT, value=0)
spi = SPI(1, sck=Pin(7), mosi=Pin(9), miso=Pin(11))

nrf = NRF24L01(spi, csn, ce, payload_size=8) # ii = 4*2 bytes
pipes = (b"\xe1\xf0\xf0\xf0\xf0", b"\xd2\xf0\xf0\xf0\xf0")

def sender():
    nrf.open_tx_pipe(pipes[0])
    nrf.open_rx_pipe(1, pipes[1])
    nrf.stop_listening()

    n = 0
    while True:
        ts = time.ticks_ms(); n += 1; m = n%10
        print("sending:", ts, m)
        try:
            nrf.send(struct.pack("ii", ts, m)) # i int size=4
        except OSError:
            pass

        time.sleep_ms(500) # delay the loop

def listener():
    nrf.open_tx_pipe(pipes[1])
    nrf.open_rx_pipe(1, pipes[0])
    nrf.start_listening()

    while True:
        if nrf.any():
            while nrf.any():
                buf = nrf.recv()
                ts, m = struct.unpack("ii", buf)
                print("received:", ts, m)

```

I hope someone finds it useful.
